File: games/lynrummy/python/geometry.py
"""
Client-side LynRummy board geometry — collision detection and
open-spot finding. The server produces logical board deltas at
dummyLoc = (0, 0); clients (this Python auto_player, the Elm UI)
decide where stacks actually sit. The referee enforces the
"no overlap, in bounds" rule at complete_turn time.

Different clients/humans arrange stacks differently (Steve's
observation: "human players have very different tidying
approaches"). This module is one style — find the first
row-major open slot.

Pure functions on state dicts as returned by the /state endpoint.
No HTTP, no mutation.
"""

# Card-layout constants. Match games/lynrummy/card_stack.go
# (CardWidth=27) and games/lynrummy/board_geometry.go (CardHeight
# = 40, CardPitch = CardWidth + 6 = 33).
CARD_WIDTH = 27
CARD_PITCH = CARD_WIDTH + 6
CARD_HEIGHT = 40

# Board region. Match the referee's bounds in
# views/lynrummy_elm.go (ValidateTurnComplete call site).
# Margin bumped 5 → 7 on 2026-04-23 (CROWDED_BOARDS) after
# BOARD_LAB captures revealed that gap=7 between stacks
# reads as overlap to humans even though the old
# gap-must-be-at-least-5 rule accepted it. Keep in sync
# with Elm's Main.Apply.refereeBounds + Go's uses of
# BoardBounds{Margin: 7}.
BOARD_MAX_WIDTH = 800
BOARD_MAX_HEIGHT = 600
BOARD_MARGIN = 7

# Where the 800x600 board div sits in the viewport. Pinned so
# that Python (which has no DOM) and Elm agree on the viewport
# coordinate of every board stack: viewport = stack.loc +
# (BOARD_VIEWPORT_LEFT, BOARD_VIEWPORT_TOP).
#
# If you change these, update `boardViewportLeft` /
# `boardViewportTop` in
# games/lynrummy/elm/src/Game/BoardGeometry.elm
# to match.
BOARD_VIEWPORT_LEFT = 300
BOARD_VIEWPORT_TOP = 100

# Placement-sweep granularity in pixels (fallback path only).
PLACE_STEP = 10

# Packing gaps — the size of the "breathing room" the agent
# leaves between adjacent stacks. Bigger than BOARD_MARGIN:
# the referee enforces the legal-minimum 5px, but a board
# packed at that margin reads as robotic AND leaves adjacent
# stacks looking ambiguous (is that one stack or two?).
#
# Tuned to roughly one card width — close enough to CARD_WIDTH
# that an empty gap reads as "a card could fit there" but not
# so wide that the board sprawls.
PACK_GAP_X = 30
PACK_GAP_Y = 30

# Anti-alignment offset. A fixed +2px nudge applied to every
# placement. Keeps output deterministic while breaking
# pixel-perfect alignment with grid multiples — enough to make
# the board not read as machine-tidied. Same inputs → same
# output, always.
ANTI_ALIGN_PX = 2

# Starting anchor when the board is empty. Hand-feel default;
# a little down-and-right from the top-left corner.
BOARD_START = (24, 24)  # (left, top)


# Preferred scan origin on a non-empty board — tuned 2026-04-23
# from BOARD_LAB human captures (Steve, Joshua, Emma). Humans
# don't land pre-moves near the (0, 0) corner; they favor a
# zone with some inset on both axes. (50, 90) is the lower-left
# edge of the observed landing cluster, chosen so the row-major
# scan hits it first before walking rightward/downward.
#
# "Toward the hand" — the hand column sits off the board to
# its left, so minimizing the upcoming hand-card drag means
# preferring LOW x among valid candidates. Scanning left-to-right
# from left=50 honors that while staying clear of the margin.
HUMAN_PREFERRED_ORIGIN = (50, 90)  # (left, top)


def stack_width(card_count):
    """Pixel width of a stack with n cards. 0 for n <= 0."""
    if card_count <= 0:
        return 0
    return CARD_WIDTH + (card_count - 1) * CARD_PITCH


def stack_rect(stack):
    """Bounding rect of a board stack (left, top, right, bottom).

    `stack` is a state-dict stack: {"loc":{"top","left"},
    "board_cards":[...]}.
    """
    left = stack["loc"]["left"]
    top = stack["loc"]["top"]
    return (
        left,
        top,
        left + stack_width(len(stack["board_cards"])),
        top + CARD_HEIGHT,
    )


def rects_overlap(a, b):
    """Axis-aligned rect overlap (exclusive edges)."""
    al, at, ar, ab = a
    bl, bt, br, bb = b
    return al < br and ar > bl and at < bb and ab > bt


def pad_rect(r, margin):
    l, t, rr, b = r
    return (l - margin, t - margin, rr + margin, b + margin)


def find_open_loc(existing, card_count):
    """Return {"top","left"} for a new stack of `card_count` cards
    that does not overlap any stack in `existing`.

    Human-biased placement (2026-04-23 update, driven by
    BOARD_LAB captures from three humans). The scan starts at
    `HUMAN_PREFERRED_ORIGIN` — a zone humans actually use —
    instead of the raw board margin. Rationale:

      - Humans don't land pre-moves at (7, 7). Captured data
        shows landing locs in the `y ≈ 60-120`, `x ≈ 40-500`
        band, never in the extreme top-left corner.
      - The hand column sits off the board to its LEFT in
        viewport frame, so pre-moves that shorten the
        hand-card's upcoming drag live toward LOW x in
        board frame. Scanning left-to-right from a
        non-corner origin hits that zone first.
      - Everything else (first-fit, PACK_GAP breathing room,
        ANTI_ALIGN nudge, crowded-fallback) stays the same.

    100% deterministic. Same board state → same placement.

    `existing` is a list of state-dict stacks.
    """
    new_w = stack_width(card_count)
    new_h = CARD_HEIGHT
    existing_rects = [stack_rect(s) for s in existing]

    # Empty board → fixed starting anchor.
    if not existing_rects:
        left, top = BOARD_START
        return _anti_align(left, top, new_w, new_h)

    # First-fit scan at packing gap. Tighter step than the
    # legal-margin fallback so the 2px anti-align offset
    # actually lands off-grid.
    step = 15
    origin_left, origin_top = HUMAN_PREFERRED_ORIGIN
    min_left = BOARD_MARGIN
    min_top = BOARD_MARGIN
    max_left = BOARD_MAX_WIDTH - new_w - BOARD_MARGIN
    max_top = BOARD_MAX_HEIGHT - new_h - BOARD_MARGIN

    # Clamp the preferred origin in case a future
    # HUMAN_PREFERRED_ORIGIN is too close to the right/bottom
    # edge for the requested stack size.
    start_left = min(max(origin_left, min_left), max_left)
    start_top = min(max(origin_top, min_top), max_top)

    top = start_top
    while top <= max_top:
        left = start_left
        while left <= max_left:
            padded = (
                left - PACK_GAP_X,
                top - PACK_GAP_Y,
                left + new_w + PACK_GAP_X,
                top + new_h + PACK_GAP_Y,
            )
            if not any(rects_overlap(padded, er) for er in existing_rects):
                return _anti_align(left, top, new_w, new_h)
            left += step
        top += step

    # Preferred-zone scan exhausted — widen to the whole
    # board (including the top-left corner) before falling
    # through to the legal-margin crowded fallback.
    top = min_top
    while top <= max_top:
        left = min_left
        while left <= max_left:
            padded = (
                left - PACK_GAP_X,
                top - PACK_GAP_Y,
                left + new_w + PACK_GAP_X,
                top + new_h + PACK_GAP_Y,
            )
            if not any(rects_overlap(padded, er) for er in existing_rects):
                return _anti_align(left, top, new_w, new_h)
            left += step
        top += step

    # Board too crowded for the packing gap — drop to legal margin.
    return _grid_sweep_open_loc(existing_rects, new_w, new_h)


def _anti_align(left, top, new_w, new_h):
    """Apply the fixed ANTI_ALIGN_PX offset, clamped to bounds."""
    jl = min(left + ANTI_ALIGN_PX, BOARD_MAX_WIDTH - new_w)
    jt = min(top + ANTI_ALIGN_PX, BOARD_MAX_HEIGHT - new_h)
    return {"top": jt, "left": jl}


def _grid_sweep_open_loc(existing_rects, new_w, new_h):
    """Deterministic row-major sweep at the legal (BOARD_MARGIN)
    padding. Only used when packed-by-clearance can't satisfy
    the human-style spacing — the board is crowded enough that
    legal-minimum is the best we can do.
    """
    top = 0
    while top + new_h <= BOARD_MAX_HEIGHT:
        left = 0
        while left + new_w <= BOARD_MAX_WIDTH:
            candidate = (
                left - BOARD_MARGIN,
                top - BOARD_MARGIN,
                left + new_w + BOARD_MARGIN,
                top + new_h + BOARD_MARGIN,
            )
            if not any(rects_overlap(candidate, er) for er in existing_rects):
                return {"top": top, "left": left}
            left += PLACE_STEP
        top += PLACE_STEP

    fallback_top = max(0, BOARD_MAX_HEIGHT - new_h)
    return {"top": fallback_top, "left": 0}


def out_of_bounds(stack):
    """True if the stack's bounding rect extends past the board."""
    l, t, r, b = stack_rect(stack)
    return l < 0 or t < 0 or r > BOARD_MAX_WIDTH or b > BOARD_MAX_HEIGHT


def loc_clears_others(loc, card_count, board, exclude_indices=()):
    """True if a stack of `card_count` cards anchored at `loc`
    fits in bounds and doesn't overlap (padded by margin) any
    stack in `board` except those whose indices are in
    `exclude_indices`. Mirrors the referee's two checks:
    bounds-in, and padded-overlap-free.
    """
    rect = (
        loc["left"],
        loc["top"],
        loc["left"] + stack_width(card_count),
        loc["top"] + CARD_HEIGHT,
    )
    if (rect[0] < 0 or rect[1] < 0 or
            rect[2] > BOARD_MAX_WIDTH or rect[3] > BOARD_MAX_HEIGHT):
        return False
    padded = pad_rect(rect, BOARD_MARGIN)
    for i, s in enumerate(board):
        if i in exclude_indices:
            continue
        if rects_overlap(padded, stack_rect(s)):
            return False
    return True


def find_violation(board):
    """Return the index of a stack that breaks the geometry rule,
    or None. Checks out-of-bounds first, then pairwise padded
    overlap (the referee's "too close" / "actual overlap" combo).

    Returns the FIRST offending stack rather than a full report —
    callers iterate: fix one, re-check, repeat until stable.
    """
    for i, s in enumerate(board):
        if out_of_bounds(s):
            return i

    rects = [stack_rect(s) for s in board]
    for i in range(len(rects)):
        padded_i = pad_rect(rects[i], BOARD_MARGIN)
        for j in range(i + 1, len(rects)):
            if rects_overlap(padded_i, rects[j]):
                # Move the later-indexed stack — it's the one
                # that was appended most recently by a trick or
                # by a growing neighbor, so relocating it
                # (rather than the settled stack) is the less
                # disruptive choice.
                return j
    return None

File: games/lynrummy/ts/src/geometry.ts
// geometry.ts — TS port of python/geometry.py.
//
// Pure functions over a board-stacks-with-locations shape. No HTTP,
// no mutation. Used by the verb→primitive translator (verbs.ts) and
// the geometry post-pass (geometry_plan.ts) to keep the agent's
// emitted plays clear of the crowding/overlap thresholds the
// referee enforces (and stricter human-feel thresholds the agent
// enforces on its own).
//
// Constants must agree with `Game.Physics.BoardGeometry` (Elm) and
// the Go server's BoardBounds. Drift between layers is silent and
// load-bearing — when changing one, update all three.

import type { Card } from "./rules/card.ts";

// --- Stack shape with position ----------------------------------------

/** Board stack carrying a position. The BFS engine itself works on
 *  bare `Card[]`; this shape is used by the verb→primitive layer and
 *  any caller that has to reason about geometry. */
export interface BoardStack {
  readonly cards: readonly Card[];
  readonly loc: Loc;
}

export interface Loc {
  readonly top: number;
  readonly left: number;
}

export interface Rect {
  readonly left: number;
  readonly top: number;
  readonly right: number;
  readonly bottom: number;
}

// --- Layout constants — match Elm + Go ---------------------------------

export const CARD_WIDTH = 27;
export const CARD_PITCH = CARD_WIDTH + 6;  // 33
export const CARD_HEIGHT = 40;

export const BOARD_MAX_WIDTH = 800;
export const BOARD_MAX_HEIGHT = 600;
export const BOARD_MARGIN = 7;

export const BOARD_VIEWPORT_LEFT = 300;
export const BOARD_VIEWPORT_TOP = 38;

export const PLACE_STEP = 10;

export const PACK_GAP_X = 30;
export const PACK_GAP_Y = 30;

export const ANTI_ALIGN_PX = 2;

/** Empty-board anchor: a little down-and-right from the top-left corner.
 *  Final placement is `(BOARD_START + ANTI_ALIGN_PX) = (26, 26)`. */
export const BOARD_START: Loc = { left: 24, top: 24 };

/** Preferred non-empty-board scan origin (Steve, 2026-04-23): humans
 *  don't land pre-moves near the (0,0) corner; they favor a left-biased
 *  zone with some inset on both axes. Toward-the-hand bias means
 *  preferring LOW x. (50, 90) is the lower-left edge of the observed
 *  human-placement cluster. */
export const HUMAN_PREFERRED_ORIGIN: Loc = { left: 50, top: 90 };

// --- Geometry primitives ----------------------------------------------

export function stackWidth(cardCount: number): number {
  if (cardCount <= 0) return 0;
  return CARD_WIDTH + (cardCount - 1) * CARD_PITCH;
}

export function stackRect(stack: BoardStack): Rect {
  const left = stack.loc.left;
  const top = stack.loc.top;
  return {
    left,
    top,
    right: left + stackWidth(stack.cards.length),
    bottom: top + CARD_HEIGHT,
  };
}

/** Axis-aligned overlap with exclusive edges. */
export function rectsOverlap(a: Rect, b: Rect): boolean {
  return a.left < b.right && a.right > b.left && a.top < b.bottom && a.bottom > b.top;
}

export function padRect(r: Rect, margin: number): Rect {
  return {
    left: r.left - margin,
    top: r.top - margin,
    right: r.right + margin,
    bottom: r.bottom + margin,
  };
}

// --- Open-location finder ---------------------------------------------

/** Apply the fixed +2px anti-align nudge, clamped to bounds. */
function antiAlign(left: number, top: number, newW: number, newH: number): Loc {
  const jl = Math.min(left + ANTI_ALIGN_PX, BOARD_MAX_WIDTH - newW);
  const jt = Math.min(top + ANTI_ALIGN_PX, BOARD_MAX_HEIGHT - newH);
  return { left: jl, top: jt };
}

/**
 * Return a `Loc` for a new `cardCount`-card stack that doesn't overlap
 * any stack in `existing`.
 *
 * Scan is COLUMN-MAJOR (outer = left, inner = top), starting from
 * `HUMAN_PREFERRED_ORIGIN`. This embodies Steve's preference order:
 * left before right, then vertical (middle-ish) before advancing
 * further right. When the top rows are packed, the agent drops DOWN
 * in the same leftward column before shifting right — rightward is
 * the last option, not the first.
 *
 * The hand column sits off the board to the LEFT, so left-biased
 * placement also shortens the next hand-card drag.
 *
 * 100% deterministic. Same board state → same placement.
 */
export function findOpenLoc(
  existing: readonly BoardStack[],
  cardCount: number,
): Loc {
  const newW = stackWidth(cardCount);
  const newH = CARD_HEIGHT;
  const existingRects = existing.map(stackRect);

  // Empty board → fixed starting anchor.
  if (existingRects.length === 0) {
    return antiAlign(BOARD_START.left, BOARD_START.top, newW, newH);
  }

  // First-fit column-major scan at packing gap. Tighter step than the
  // legal-margin fallback so the 2px anti-align offset actually lands
  // off-grid.
  const step = 15;
  const minLeft = BOARD_MARGIN;
  const minTop = BOARD_MARGIN;
  const maxLeft = BOARD_MAX_WIDTH - newW - BOARD_MARGIN;
  const maxTop = BOARD_MAX_HEIGHT - newH - BOARD_MARGIN;

  // Clamp the preferred origin in case a future HUMAN_PREFERRED_ORIGIN
  // is too close to the right/bottom edge for the requested stack size.
  const startLeft = Math.min(Math.max(HUMAN_PREFERRED_ORIGIN.left, minLeft), maxLeft);
  const startTop = Math.min(Math.max(HUMAN_PREFERRED_ORIGIN.top, minTop), maxTop);

  const clears = (left: number, top: number): boolean => {
    const padded: Rect = {
      left: left - PACK_GAP_X,
      top: top - PACK_GAP_Y,
      right: left + newW + PACK_GAP_X,
      bottom: top + newH + PACK_GAP_Y,
    };
    for (const er of existingRects) if (rectsOverlap(padded, er)) return false;
    return true;
  };

  // Phase 1: scan from the preferred origin. Column-major so a packed
  // top row drops us downward in-place rather than pushing rightward.
  for (let left = startLeft; left <= maxLeft; left += step) {
    for (let top = startTop; top <= maxTop; top += step) {
      if (clears(left, top)) return antiAlign(left, top, newW, newH);
    }
  }

  // Phase 2: widen to the whole board, still column-major, before
  // falling through to the legal-margin crowded fallback.
  for (let left = minLeft; left <= maxLeft; left += step) {
    for (let top = minTop; top <= maxTop; top += step) {
      if (clears(left, top)) return antiAlign(left, top, newW, newH);
    }
  }

  // Board too crowded for the packing gap — drop to legal margin.
  return gridSweepOpenLoc(existingRects, newW, newH);
}

/** Deterministic row-major sweep at the legal (BOARD_MARGIN) padding.
 *  Fallback when packed-by-clearance can't satisfy the human-style
 *  spacing — the board is crowded enough that legal-minimum is the
 *  best we can do. */
function gridSweepOpenLoc(
  existingRects: readonly Rect[],
  newW: number,
  newH: number,
): Loc {
  for (let top = 0; top + newH <= BOARD_MAX_HEIGHT; top += PLACE_STEP) {
    for (let left = 0; left + newW <= BOARD_MAX_WIDTH; left += PLACE_STEP) {
      const candidate: Rect = {
        left: left - BOARD_MARGIN,
        top: top - BOARD_MARGIN,
        right: left + newW + BOARD_MARGIN,
        bottom: top + newH + BOARD_MARGIN,
      };
      let clears = true;
      for (const er of existingRects) {
        if (rectsOverlap(candidate, er)) { clears = false; break; }
      }
      if (clears) return { top, left };
    }
  }
  // Board fully crowded — fall back to bottom-left corner.
  const fallbackTop = Math.max(0, BOARD_MAX_HEIGHT - newH);
  return { top: fallbackTop, left: 0 };
}

// --- Bounds + violation checks ----------------------------------------

export function outOfBounds(stack: BoardStack): boolean {
  const r = stackRect(stack);
  return r.left < 0 || r.top < 0 || r.right > BOARD_MAX_WIDTH || r.bottom > BOARD_MAX_HEIGHT;
}

/** True iff a stack of `cardCount` cards anchored at `loc` fits in
 *  bounds and doesn't overlap (padded by BOARD_MARGIN) any stack in
 *  `board` except those whose indices are in `excludeIndices`.
 *  Mirrors the referee's two checks. */
export function locClearsOthers(
  loc: Loc,
  cardCount: number,
  board: readonly BoardStack[],
  excludeIndices: ReadonlySet<number> = new Set(),
): boolean {
  const rect: Rect = {
    left: loc.left,
    top: loc.top,
    right: loc.left + stackWidth(cardCount),
    bottom: loc.top + CARD_HEIGHT,
  };
  if (rect.left < 0 || rect.top < 0
      || rect.right > BOARD_MAX_WIDTH || rect.bottom > BOARD_MAX_HEIGHT) {
    return false;
  }
  const padded = padRect(rect, BOARD_MARGIN);
  for (let i = 0; i < board.length; i++) {
    if (excludeIndices.has(i)) continue;
    if (rectsOverlap(padded, stackRect(board[i]!))) return false;
  }
  return true;
}

/** Return the index of a stack that breaks the geometry rule, or null.
 *  Checks out-of-bounds first, then pairwise padded overlap.
 *
 *  Returns the FIRST offending stack rather than a full report —
 *  callers iterate: fix one, re-check, repeat until stable. When two
 *  stacks overlap, the LATER-indexed one is reported (it's the one
 *  that was appended most recently by a trick or by a growing
 *  neighbor; relocating it is less disruptive). */
export function findViolation(board: readonly BoardStack[]): number | null {
  for (let i = 0; i < board.length; i++) {
    if (outOfBounds(board[i]!)) return i;
  }
  const rects = board.map(stackRect);
  for (let i = 0; i < rects.length; i++) {
    const paddedI = padRect(rects[i]!, BOARD_MARGIN);
    for (let j = i + 1; j < rects.length; j++) {
      if (rectsOverlap(paddedI, rects[j]!)) return j;
    }
  }
  return null;
}
